lom/src/plugins/plugins_common/plugins_helper.go

```diff
@@ -164,70 +184,163 @@ func (periodicDetectionPluginUtil *PeriodicDetectionPluginUtil) Init(pluginName
+                    lomcommon.LogError(fmt.Sprintf("Detection logic for plugin %s has timedout", periodicDetectionPluginUtil.PluginName))
                 }
             }
 
-        case <-periodicDetectionPluginUtil.shutDownChannel:
+        case resp := <-periodicDetectionPluginUtil.responseChannel:
+            return resp
+
+        case <-periodicDetectionPluginUtil.ctx.Done():
             /* Shutdown stops the periodic detection */
             lomcommon.LogInfo(fmt.Sprintf("Aborting Request for (%s)", periodicDetectionPluginUtil.PluginName))
             responseData := GetResponse(request, "", "", ResultCodeAborted, ResultStringFailure)
-            periodicDetectionPluginUtil.requestAborted.Store(true)
             return responseData
         }
     }
 }
 
-/* Shutdown that aborts the request. It also cleans up plugin defined cleanUp at the end */
-func (periodicDetectionPluginUtil *PeriodicDetectionPluginUtil) Shutdown() error {
-    lomcommon.LogInfo(fmt.Sprintf("Shutdown called for plugin (%s)", periodicDetectionPluginUtil.PluginName))
-    close(periodicDetectionPluginUtil.shutDownChannel)
-    startTime := time.Now()
+/* Publishes heartbeat after performing validations */
+func (periodicDetectionPluginUtil *PeriodicDetectionPluginUtil) publishHeartBeat(hbchan chan PluginHeartBeat) {
+
+    numConsecutiveErrors := periodicDetectionPluginUtil.numOfConsecutiveErrors.Load()
+    periodicDetectionPluginUtil.detectionRunInfo.mutex.Lock()
+    durationOfLatestRunInSecs := periodicDetectionPluginUtil.detectionRunInfo.durationOfLatestRunInSeconds
+    var currentRunStartTimeInUtc time.Time
+    if periodicDetectionPluginUtil.detectionRunInfo.currentRunStartTimeInUtc != nil {
+        currentRunStartTimeInUtc = *periodicDetectionPluginUtil.detectionRunInfo.currentRunStartTimeInUtc
+    }
+    periodicDetectionPluginUtil.detectionRunInfo.mutex.Unlock()
+
+    if numConsecutiveErrors >= uint64(lomcommon.GetConfigMgr().GetGlobalCfgInt(min_err_cnt_to_skip_hb_key)) {
+        lomcommon.LogError(fmt.Sprintf("Skipping heartbeat for %s. numConsecutiveErrors %d", periodicDetectionPluginUtil.PluginName, numConsecutiveErrors))
+        return
+    } else if durationOfLatestRunInSecs > int64(periodicDetectionPluginUtil.requestFrequencyInSecs) {
+        lomcommon.LogError(fmt.Sprintf("Skipping heartbeat for %s. DurationOfLatestRunInSecs %d", periodicDetectionPluginUtil.PluginName, durationOfLatestRunInSecs))
+        return
+    } else if !currentRunStartTimeInUtc.IsZero() { /* Indicates a request is running currently */
+        durationTillNow := int64(time.Since(currentRunStartTimeInUtc).Seconds())
+        if durationTillNow > int64(periodicDetectionPluginUtil.requestFrequencyInSecs) {
+            lomcommon.LogError(fmt.Sprintf("Skipping heartbeat for %s. Duration of current execution in secs %d", periodicDetectionPluginUtil.PluginName, durationTillNow))
+            return
+        }
+    }
+
```

Comment:
handleRequest1 error , long time
handleRequest2 no error , long time
handleRequest3 error , long time

...

is there any chance heartbeats are skipped in wrost case ?

Reply:
heartbeat will get skipped after handleRequest1 until the the request completes on time. In the above case, hb wont be sent after handleRequest1.